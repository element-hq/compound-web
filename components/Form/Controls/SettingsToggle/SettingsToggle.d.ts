import { ComponentProps, PropsWithChildren, default as React } from '../../../../../node_modules/react';
import { ToggleInput } from '../Toggle';
type SettingsToggleProps = {
    /**
     * The CSS class name for the containing field.
     */
    className?: string;
    /**
     * The field name.
     */
    name: string;
    label: string;
    /**
     * Optional help text to display below the setting.
     */
    helpMessage?: string;
    /**
     * Optional help text to display below the setting.
     */
    disabledMessage?: string;
} & Omit<ComponentProps<typeof ToggleInput>, "type">;
/**
 * A toggle component.
 */
export declare const SettingsToggleInput: React.ForwardRefExoticComponent<Omit<PropsWithChildren<SettingsToggleProps>, "ref"> & React.RefAttributes<HTMLInputElement>>;
/**
 * A styled checkbox input wrapped in a `Control` component, for use in Radix forms.
 */
export declare const SettingsToggleControl: React.ForwardRefExoticComponent<Omit<Omit<PropsWithChildren<SettingsToggleProps>, "ref"> & React.RefAttributes<HTMLInputElement>, "ref"> & React.RefAttributes<HTMLInputElement>>;
export {};
//# sourceMappingURL=SettingsToggle.d.ts.map