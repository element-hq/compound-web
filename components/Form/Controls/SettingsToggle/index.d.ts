export { SettingsToggleControl, SettingsToggleInput } from './SettingsToggle';
//# sourceMappingURL=index.d.ts.map