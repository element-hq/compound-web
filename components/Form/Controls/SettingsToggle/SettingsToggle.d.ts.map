{"version":3,"file":"SettingsToggle.d.ts","sourceRoot":"","sources":["../../../../../src/components/Form/Controls/SettingsToggle/SettingsToggle.tsx"],"names":[],"mappings":"AAOA,OAAO,EACL,cAAc,EAGd,iBAAiB,EAClB,MAAM,OAAO,CAAC;AACf,OAAO,EAAE,WAAW,EAAE,MAAM,WAAW,CAAC;AAIxC,OAAO,KAAK,MAAM,OAAO,CAAC;AAG1B,KAAK,mBAAmB,GAAG;IACzB;;OAEG;IACH,SAAS,CAAC,EAAE,MAAM,CAAC;IACnB;;OAEG;IACH,IAAI,EAAE,MAAM,CAAC;IACb,KAAK,EAAE,MAAM,CAAC;IACd;;OAEG;IACH,WAAW,CAAC,EAAE,MAAM,CAAC;IACrB;;OAEG;IACH,eAAe,CAAC,EAAE,MAAM,CAAC;CAC1B,GAAG,IAAI,CAAC,cAAc,CAAC,OAAO,WAAW,CAAC,EAAE,MAAM,CAAC,CAAC;AAErD;;GAEG;AACH,eAAO,MAAM,mBAAmB,8HAqB9B,CAAC;AAEH;;GAEG;AACH,eAAO,MAAM,qBAAqB,mLAShC,CAAC"}