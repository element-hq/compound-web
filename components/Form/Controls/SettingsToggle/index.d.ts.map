{"version":3,"file":"index.d.ts","sourceRoot":"","sources":["../../../../../src/components/Form/Controls/SettingsToggle/index.ts"],"names":[],"mappings":"AASA,OAAO,EAAE,qBAAqB,EAAE,mBAAmB,EAAE,MAAM,kBAAkB,CAAC"}