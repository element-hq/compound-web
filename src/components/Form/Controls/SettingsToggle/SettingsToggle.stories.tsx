/*
Copyright 2025 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import { Meta, StoryObj } from "@storybook/react";

import { SettingsToggleInput } from "./SettingsToggle";
import { Form } from "@radix-ui/react-form";
import { TooltipProvider } from "../../../Tooltip/TooltipProvider";
import React from "react";

const meta = {
  title: "Form/Controls/SettingsToggle",
  component: SettingsToggleInput,
  decorators: [
    (Story) => (
      <TooltipProvider>
        <Form>
          <Story />
        </Form>
      </TooltipProvider>
    ),
  ],
  tags: ["autodocs"],
  argTypes: {},
  args: {
    label: "Setting name",
    helpMessage: "This is some help text to explain what this setting does.",
  },
} satisfies Meta<typeof SettingsToggleInput>;
export default meta;

type Story = StoryObj<typeof meta>;

export const Unchecked: Story = {
  args: {
    defaultChecked: false,
  },
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/design/qTWRfItpO3RdCjnTKPu4mL/Settings?node-id=97-15874",
    },
  },
};

export const Checked: Story = {
  args: {
    defaultChecked: true,
  },
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/design/qTWRfItpO3RdCjnTKPu4mL/Settings?node-id=97-15874",
    },
  },
};

export const DisabledUnchecked: Story = {
  args: {
    disabled: true,
    disabledMessage: "This feature isn't available to you right now.",
  },
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/design/qTWRfItpO3RdCjnTKPu4mL/Settings?node-id=149-38574",
    },
  },
};

export const DisabledChecked: Story = {
  args: {
    disabled: true,
    defaultChecked: true,
    disabledMessage: "This feature cannot be disabled.",
  },
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/design/qTWRfItpO3RdCjnTKPu4mL/Settings?node-id=149-38574",
    },
  },
};