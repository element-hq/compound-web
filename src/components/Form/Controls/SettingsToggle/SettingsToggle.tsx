/*
Copyright 2025 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import {
  ComponentProps,
  ComponentRef,
  forwardRef,
  PropsWithChildren,
} from "react";
import { ToggleInput } from "../Toggle";
import { Label } from "../../Label";
import { HelpMessage } from "../../Message";
import { InlineField } from "../../InlineField";
import React from "react";
import { Control } from "@radix-ui/react-form";

type SettingsToggleProps = {
  /**
   * The CSS class name for the containing field.
   */
  className?: string;
  /**
   * The field name.
   */
  name: string;
  label: string;
  /**
   * Optional help text to display below the setting.
   */
  helpMessage?: string;
  /**
   * Optional help text to display below the setting.
   */
  disabledMessage?: string;
} & Omit<ComponentProps<typeof ToggleInput>, "type">;

/**
 * A toggle component.
 */
export const SettingsToggleInput = forwardRef<
  HTMLInputElement,
  PropsWithChildren<SettingsToggleProps>
>(function Toggle(
  { className, label, helpMessage, disabledMessage, name, ...props },
  ref,
) {
  const content = (
    <InlineField
      className={className}
      name={name}
      control={<ToggleInput ref={ref} {...props} />}
    >
      <Label>{label}</Label>
      {helpMessage && <HelpMessage>{helpMessage}</HelpMessage>}
      {disabledMessage && props.disabled && (
        <HelpMessage>{disabledMessage}</HelpMessage>
      )}
    </InlineField>
  );
  return content;
});

/**
 * A styled checkbox input wrapped in a `Control` component, for use in Radix forms.
 */
export const SettingsToggleControl = forwardRef<
  ComponentRef<typeof SettingsToggleInput>,
  ComponentProps<typeof SettingsToggleInput>
>(function ToggleControl(props, ref) {
  return (
    <Control asChild>
      <SettingsToggleInput ref={ref} {...props} />
    </Control>
  );
});
