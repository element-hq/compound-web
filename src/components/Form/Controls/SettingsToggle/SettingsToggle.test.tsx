/*
Copyright 2025 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import { describe, it, expect } from "vitest";
import { render } from "@testing-library/react";
import React from "react";
import { SettingsToggleInput } from "./SettingsToggle";
import { Form } from "@radix-ui/react-form";

describe("Toggle", () => {
  it("renders", () => {
    const { asFragment } = render(<Form><SettingsToggleInput /></Form>);
    expect(asFragment()).toMatchSnapshot();
  });
});
