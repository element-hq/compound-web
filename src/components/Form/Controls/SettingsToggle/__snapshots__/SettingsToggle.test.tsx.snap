// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Toggle > renders 1`] = `
<DocumentFragment>
  <form>
    <div
      class="_inline-field_af3470"
    >
      <div
        class="_inline-field-control_af3470"
      >
        <div
          class="_container_d9763f"
        >
          <input
            class="_input_d9763f"
            id="«r0»"
            role="switch"
            type="checkbox"
          />
          <div
            class="_ui_d9763f"
          />
        </div>
      </div>
      <div
        class="_inline-field-body_af3470"
      >
        <label
          class="_label_af3470"
          for="«r0»"
        >
          Simple setting
        </label>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`Toggle > renders with a help message 1`] = `
<DocumentFragment>
  <form>
    <div
      class="_inline-field_af3470"
    >
      <div
        class="_inline-field-control_af3470"
      >
        <div
          class="_container_d9763f"
        >
          <input
            class="_input_d9763f"
            id="«r2»"
            role="switch"
            type="checkbox"
          />
          <div
            class="_ui_d9763f"
          />
        </div>
      </div>
      <div
        class="_inline-field-body_af3470"
      >
        <label
          class="_label_af3470"
          for="«r2»"
        >
          Simple setting
        </label>
        <span
          class="_message_af3470 _help-message_af3470"
          id="radix-«r4»"
        >
          A simple setting to control things
        </span>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`renders with a disabled message (disabled=false) 1`] = `
<DocumentFragment>
  <form>
    <div
      class="_inline-field_af3470"
    >
      <div
        class="_inline-field-control_af3470"
      >
        <div
          class="_container_d9763f"
        >
          <input
            class="_input_d9763f"
            id="«r9»"
            role="switch"
            type="checkbox"
          />
          <div
            class="_ui_d9763f"
          />
        </div>
      </div>
      <div
        class="_inline-field-body_af3470"
      >
        <label
          class="_label_af3470"
          for="«r9»"
        >
          Simple setting
        </label>
        <span
          class="_message_af3470 _help-message_af3470"
          id="radix-«rb»"
        >
          A simple setting to control things
        </span>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`renders with a disabled message (disabled=true) 1`] = `
<DocumentFragment>
  <form>
    <div
      class="_inline-field_af3470"
    >
      <div
        class="_inline-field-control_af3470"
      >
        <div
          class="_container_d9763f"
        >
          <input
            class="_input_d9763f"
            disabled=""
            id="«r5»"
            role="switch"
            type="checkbox"
          />
          <div
            class="_ui_d9763f"
          />
        </div>
      </div>
      <div
        class="_inline-field-body_af3470"
      >
        <label
          class="_label_af3470"
          for="«r5»"
        >
          Simple setting
        </label>
        <span
          class="_message_af3470 _help-message_af3470"
          id="radix-«r7»"
        >
          A simple setting to control things
        </span>
        <span
          class="_message_af3470 _help-message_af3470"
          id="radix-«r8»"
        >
          This is disabled.
        </span>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
